commons/com.b2international.index/src/com/b2international/index/es/EsDocumentSearcher.java

```diff
@@ -532,5 +541,41 @@ public int resultWindow() {
 	public int maxTermsCount() {
 		return maxTermsCount;
 	}
+	
+	@Override
+	public <T> Hits<T> knn(Knn<T> knn) throws IOException {
+		if (admin.es8Client() == null) {
+			throw new BadRequestException("Approximate knn search is only available in Elastiscearch 8 clusters. The currently connected ES cluster version is %s.", admin.client().version());
+		}
+		
+		final ElasticsearchClient client = admin.es8Client().client();
+		final DocumentMapping mapping = admin.mappings().getMapping(knn.getFrom());
+		
+		Es8QueryBuilder q = new Es8QueryBuilder(mapping, admin.settings(), admin.log());
+		co.elastic.clients.elasticsearch._types.query_dsl.Query esQuery = q.build(knn.getFilter());
+		
+		// TODO consider adding support for double primitive lists
+		FloatIterator it = knn.getQueryVector().iterator();
+		final List<Double> queryVector = new ArrayList<>(knn.getQueryVector().size());
+		while (it.hasNext()) {
+			queryVector.add((double) it.next());
+		}
```

Comment:
This is a bit strange on the ES Java Client's part, as the internal representation is certainly limited to 32 bits in precision, as mentioned in [Dense vector field type](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html) and [Functions for vector fields](https://www.elastic.co/guide/en/elasticsearch/reference/master/query-dsl-script-score-query.html#vector-functions):

> The dense_vector field type stores dense vectors of float values.

> `doc[<field>].vectorValue` – returns a vector’s value as an array of floats

Doubles are most likely used to overcome some JSON serialization limitations. I don't think our Knn search vector should adopt this type and let users think that higher precision can be used than what is actually available.

Reply:
Yes, that's why I have used FloatList instead of introducing doubles. In our API we will index an array of float values and will support searching using an array of float values, no doubles.